src/image_bios.cpp

```diff
@@ -0,0 +1,196 @@
+    }
+    catch (const std::exception& e)
+    {
+        throw FwupdateError("Unable to %s SPI driver! %s", actionStr.c_str(),
+                            e.what());
+    }
+
+    constexpr auto maxDriverBindingChecks = 100;
+    constexpr auto delayBetweenChecksMS = 20;
+    for (int i = 0; i < maxDriverBindingChecks; ++i)
+    {
+        if (isSPIDriverBound() == action)
+        {
+            // Operation completed.
+            return;
+        }
+
+        std::this_thread::sleep_for(
+            std::chrono::milliseconds(delayBetweenChecksMS));
+    }
+
+    throw FwupdateError("The %s SPI driver timed out!",
+                        action ? "binding" : "unbinding");
+}
+
+static void bindSPIDriver()
+{
+    if (!isSPIDriverBound())
+    {
+        bindOrUnbindSPIDriver(true);
+    }
+}
+
+static void unbindSPIDriver()
+{
+    if (isSPIDriverBound())
+    {
+        bindOrUnbindSPIDriver(false);
+    }
+}
+
+/**
+ * @brief Take control of specified GPIO line and set its direction to the
+ *        output mode and wanted value.
+ *
+ * @param name     - GPIO line name
+ * @param value    - Wanted value
+ * @param gpioLine - Reference of GPIO line object
+ */
+void setGPIOOutput(const std::string& name, const int value,
+                   gpiod::line& gpioLine)
+{
+    gpioLine = gpiod::find_line(name);
+    if (!gpioLine)
+    {
+        throw FwupdateError("GPIO line %s not found!", name.c_str());
+    }
+
+    try
+    {
+        gpioLine.request({gpioOwner, gpiod::line_request::DIRECTION_OUTPUT, 0},
+                         value);
+    }
+    catch (const std::exception& e)
+    {
+        throw FwupdateError("Unable to set GPIO %s! %s", name.c_str(),
+                            e.what());
+    }
+}
+
+/**
+ * @brief Release GPIO line and restore default state.
+ */
+static void releaseGPIO(gpiod::line& gpioLine)
+{
+    if (gpioLine)
+    {
+        try
+        {
+            gpioLine.release();
+            gpioLine.request(
+                {gpioOwner, gpiod::line_request::DIRECTION_INPUT, 0});
+        }
+        catch (const std::exception& e)
+        {
+            throw FwupdateError("Unable to release GPIO %s! %s",
+                                gpioLine.name().c_str(), e.what());
+        }
+
+        gpioLine.reset();
+    }
+}
+
+void BIOSUpdater::lock()
+{
+    if (!files.empty())
+    {
+        Tracer tracer("Shutting down PCH");
+
+        // This makes the BMC to reinit the driver which might be attached with
+        // wrong data on the hosts equipped with em100.
+        unbindSPIDriver();
+
+        if (isChassisOn())
+        {
+            throw FwupdateError(
+                "The host is running now, operation cancelled!");
+        }
+
+        setGPIOOutput(gpioNamePCHPower, 0, gpioPCHPower);
+        setGPIOOutput(gpioNameBIOSSel, 1, gpioBIOSSel);
+        bindSPIDriver();
+
+        locked = true;
+        tracer.done();
+    }
+}
+
+void BIOSUpdater::unlock()
+{
+    if (locked)
```

Comment:
Doesn't look like a lock actually, provided that it doesn't affect any return value or exception throwing.
More like a `bound` flag.
Is flash locking done elsewhere? I mean, what happens if a second copy of `fwupdate` is ran?

Reply:
Yeah, you are right. I missed this aspect.
And two copy of `fwupdate` would disturb each other.
So, I rewrote the code and now use the PCH Power pin as a lock.